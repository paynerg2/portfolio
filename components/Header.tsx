import {
    Stack,
    Heading,
    Flex,
    useDisclosure,
    useColorModeValue,
    IconButton,
    Drawer,
    DrawerOverlay,
    DrawerContent,
    DrawerCloseButton,
    DrawerHeader,
    DrawerBody,
    Button,
    Icon,
    Box,
} from '@chakra-ui/react';
import { HamburgerIcon } from '@chakra-ui/icons';
import { useRef } from 'react';
import { sectionNames } from '../utils/constants';
import { scrollTo as scroll } from '../utils/scrollTo';
import { FaGithub, FaLinkedin } from 'react-icons/fa';

const Header = () => {
    const headerColor = useColorModeValue('gray.50', 'gray.900');
    const headerTextColor = useColorModeValue('gray.900', 'white');
    const { isOpen, onOpen, onClose } = useDisclosure();
    const menuButtonRef = useRef(null);

    const handleToggle = () => (isOpen ? onClose() : onOpen());

    const scrollTo = (id: string) => {
        scroll({ id, before: onClose });
    };

    // TODO: Refactor into separate component that handles its own display logic
    const getNavLinks = (viewport: string) => {
        const direction = viewport === 'desktop' ? 'row' : 'column';
        const display = viewport === 'desktop' ? 'none' : 'flex';
        return (
            <Stack
                direction={{ base: 'column', md: direction }}
                display={{ base: display, md: 'flex' }}
                width={{ base: 'fill', md: 'fill' }}
                align="center"
                justify="center"
                flexGrow={1}
                spacing={10}
            >
                <Button
                    onClick={() => scrollTo(sectionNames.projects)}
                    colorScheme="white"
                    variant="link"
                >
                    Projects
                </Button>
                <Button
                    onClick={() => scrollTo(sectionNames.skills)}
                    colorScheme="white"
                    variant="link"
                >
                    Skills
                </Button>
                <Button
                    onClick={() => scrollTo(sectionNames.about)}
                    colorScheme="white"
                    variant="link"
                >
                    About Me
                </Button>
                <Button
                    onClick={() => scrollTo(sectionNames.contact)}
                    colorScheme="white"
                    variant="link"
                >
                    Contact
                </Button>
                <Button
                    as="a"
                    href="https://resume.creddle.io/resume/jltrgnkt9gg"
                    target="_blank"
                    rel="noreferrer noopener"
                    colorScheme="white"
                    variant="outline"
                >
                    Resume
                </Button>
            </Stack>
        );
    };

    return (
        <Flex
            as="nav"
            wrap="wrap"
            w="100vw"
            padding={6}
            mb={12}
            bg={headerColor}
            color={headerTextColor}
        >
            {/* Mobile Nav */}
            <IconButton
                ref={menuButtonRef}
                onClick={handleToggle}
                display={{ base: 'block', md: 'none' }}
                aria-label="Menu"
                icon={<HamburgerIcon boxSize={8} />}
            />
            <Drawer
                isOpen={isOpen}
                placement="left"
                onClose={onClose}
                finalFocusRef={menuButtonRef}
            >
                <DrawerOverlay />
                <DrawerContent>
                    <DrawerHeader>
                        <DrawerCloseButton />
                        <Heading>Sections</Heading>
                    </DrawerHeader>

                    <DrawerBody>{getNavLinks('mobile')}</DrawerBody>
                </DrawerContent>
            </Drawer>

            {/* Desktop Nav */}
            {getNavLinks('desktop')}
        </Flex>
    );
};

export default Header;
