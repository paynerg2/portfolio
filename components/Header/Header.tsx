import {
    Stack,
    Heading,
    Flex,
    useDisclosure,
    IconButton,
    Drawer,
    DrawerOverlay,
    DrawerContent,
    DrawerCloseButton,
    DrawerHeader,
    DrawerBody,
    Button,
} from '@chakra-ui/react';
import { HamburgerIcon } from '@chakra-ui/icons';
import { useRef } from 'react';
import { sectionNames } from '../../utils/constants';
import { scrollTo as scroll } from '../../utils/scrollTo';
import { useRouter } from 'next/router';

const Header = () => {
    const router = useRouter();
    const { isOpen, onOpen, onClose } = useDisclosure();
    const menuButtonRef = useRef(null);

    const handleToggle = () => (isOpen ? onClose() : onOpen());

    const scrollTo = (id: string) => {
        scroll({ id, before: onClose });
    };

    const handleNav = (id: string) => {
        if (router.pathname !== '/') {
            router.push('/');
        }
        scrollTo(id);
    };

    // TODO: Refactor into separate component that handles its own display logic
    const getNavLinks = (viewport: string) => {
        const direction = viewport === 'desktop' ? 'row' : 'column';
        const display = viewport === 'desktop' ? 'none' : 'flex';
        return (
            <Stack
                direction={{ base: 'column', md: direction }}
                display={{ base: display, md: 'flex' }}
                width={{ base: 'fill', md: 'fill' }}
                align="center"
                justify="center"
                flexGrow={1}
                spacing={10}
            >
                <Button
                    onClick={() => scrollTo(sectionNames.projects)}
                    colorScheme="white"
                    variant="link"
                >
                    Projects
                </Button>
                <Button
                    onClick={() => handleNav(sectionNames.about)}
                    colorScheme="white"
                    variant="link"
                >
                    About Me
                </Button>
                <Button
                    onClick={() => handleNav(sectionNames.contact)}
                    colorScheme="white"
                    variant="link"
                >
                    Contact
                </Button>
                <Button
                    as="a"
                    href="https://resume.creddle.io/resume/jltrgnkt9gg"
                    target="_blank"
                    rel="noreferrer noopener"
                    colorScheme="white"
                    variant="outline"
                >
                    Resume
                </Button>
            </Stack>
        );
    };

    return (
        <Flex as="nav" wrap="wrap" w="100vw" padding={6} mb={12} bg="gray.900" color="white">
            {/* Mobile Nav */}
            <IconButton
                ref={menuButtonRef}
                onClick={handleToggle}
                display={{ base: 'block', md: 'none' }}
                aria-label="Menu"
                icon={<HamburgerIcon boxSize={8} />}
            />
            <Drawer
                isOpen={isOpen}
                placement="left"
                onClose={onClose}
                finalFocusRef={menuButtonRef}
            >
                <DrawerOverlay />
                <DrawerContent>
                    <DrawerHeader>
                        <DrawerCloseButton />
                        <Heading>Sections</Heading>
                    </DrawerHeader>

                    <DrawerBody>{getNavLinks('mobile')}</DrawerBody>
                </DrawerContent>
            </Drawer>

            {/* Desktop Nav */}
            {getNavLinks('desktop')}
        </Flex>
    );
};

export default Header;
