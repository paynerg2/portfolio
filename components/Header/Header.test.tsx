import { render } from '../../test/test-utils';
import Header from './Header';

describe('Header Component', () => {
    describe('Desktop Window Size', () => {
        it('renders the component', () => {
            const component = render(<Header />);
            expect(component).toMatchSnapshot();
        });
    });

    describe('Mobile Window Size', () => {
        Object.defineProperty(window, 'innerHeight', {
            writable: true,
            configurable: true,
            value: 150,
        });
        window.dispatchEvent(new Event('resize'));

        it('renders the component', () => {
            expect(window.innerHeight).toBe(150);
            const component = render(<Header />);
            expect(component).toMatchSnapshot();
        });
    });
});
