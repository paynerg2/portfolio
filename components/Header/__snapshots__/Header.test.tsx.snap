// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component Desktop Window Size renders the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-dark"
  >
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
    <div>
      <nav
        class="css-h56wkw"
      >
        <button
          aria-label="Menu"
          class="chakra-button css-s2kmp4"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-6tzxjd"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M 3 5 A 1.0001 1.0001 0 1 0 3 7 L 21 7 A 1.0001 1.0001 0 1 0 21 5 L 3 5 z M 3 11 A 1.0001 1.0001 0 1 0 3 13 L 21 13 A 1.0001 1.0001 0 1 0 21 11 L 3 11 z M 3 17 A 1.0001 1.0001 0 1 0 3 19 L 21 19 A 1.0001 1.0001 0 1 0 21 17 L 3 17 z"
              fill="currentColor"
            />
          </svg>
        </button>
        <div
          class="chakra-stack css-1wtjd1r"
        >
          <button
            class="chakra-button css-itouur"
            type="button"
          >
            Projects
          </button>
          <button
            class="chakra-button css-itouur"
            type="button"
          >
            About Me
          </button>
          <button
            class="chakra-button css-itouur"
            type="button"
          >
            Contact
          </button>
          <a
            class="chakra-button css-1s4lge1"
            href="https://resume.creddle.io/resume/jltrgnkt9gg"
            rel="noreferrer noopener"
            target="_blank"
          >
            Resume
          </a>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="css-h56wkw"
    >
      <button
        aria-label="Menu"
        class="chakra-button css-s2kmp4"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-6tzxjd"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M 3 5 A 1.0001 1.0001 0 1 0 3 7 L 21 7 A 1.0001 1.0001 0 1 0 21 5 L 3 5 z M 3 11 A 1.0001 1.0001 0 1 0 3 13 L 21 13 A 1.0001 1.0001 0 1 0 21 11 L 3 11 z M 3 17 A 1.0001 1.0001 0 1 0 3 19 L 21 19 A 1.0001 1.0001 0 1 0 21 17 L 3 17 z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="chakra-stack css-1wtjd1r"
      >
        <button
          class="chakra-button css-itouur"
          type="button"
        >
          Projects
        </button>
        <button
          class="chakra-button css-itouur"
          type="button"
        >
          About Me
        </button>
        <button
          class="chakra-button css-itouur"
          type="button"
        >
          Contact
        </button>
        <a
          class="chakra-button css-1s4lge1"
          href="https://resume.creddle.io/resume/jltrgnkt9gg"
          rel="noreferrer noopener"
          target="_blank"
        >
          Resume
        </a>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Header Component Mobile Window Size renders the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-dark"
  >
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
    <div>
      <nav
        class="css-h56wkw"
      >
        <button
          aria-label="Menu"
          class="chakra-button css-s2kmp4"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-6tzxjd"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M 3 5 A 1.0001 1.0001 0 1 0 3 7 L 21 7 A 1.0001 1.0001 0 1 0 21 5 L 3 5 z M 3 11 A 1.0001 1.0001 0 1 0 3 13 L 21 13 A 1.0001 1.0001 0 1 0 21 11 L 3 11 z M 3 17 A 1.0001 1.0001 0 1 0 3 19 L 21 19 A 1.0001 1.0001 0 1 0 21 17 L 3 17 z"
              fill="currentColor"
            />
          </svg>
        </button>
        <div
          class="chakra-stack css-1wtjd1r"
        >
          <button
            class="chakra-button css-itouur"
            type="button"
          >
            Projects
          </button>
          <button
            class="chakra-button css-itouur"
            type="button"
          >
            About Me
          </button>
          <button
            class="chakra-button css-itouur"
            type="button"
          >
            Contact
          </button>
          <a
            class="chakra-button css-1s4lge1"
            href="https://resume.creddle.io/resume/jltrgnkt9gg"
            rel="noreferrer noopener"
            target="_blank"
          >
            Resume
          </a>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="css-h56wkw"
    >
      <button
        aria-label="Menu"
        class="chakra-button css-s2kmp4"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-6tzxjd"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M 3 5 A 1.0001 1.0001 0 1 0 3 7 L 21 7 A 1.0001 1.0001 0 1 0 21 5 L 3 5 z M 3 11 A 1.0001 1.0001 0 1 0 3 13 L 21 13 A 1.0001 1.0001 0 1 0 21 11 L 3 11 z M 3 17 A 1.0001 1.0001 0 1 0 3 19 L 21 19 A 1.0001 1.0001 0 1 0 21 17 L 3 17 z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="chakra-stack css-1wtjd1r"
      >
        <button
          class="chakra-button css-itouur"
          type="button"
        >
          Projects
        </button>
        <button
          class="chakra-button css-itouur"
          type="button"
        >
          About Me
        </button>
        <button
          class="chakra-button css-itouur"
          type="button"
        >
          Contact
        </button>
        <a
          class="chakra-button css-1s4lge1"
          href="https://resume.creddle.io/resume/jltrgnkt9gg"
          rel="noreferrer noopener"
          target="_blank"
        >
          Resume
        </a>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
